### PR TITLE
add memory access patch for QEMU v2.4.0.1

### PR DESCRIPTION
**Pull request by Wenzel**
Hi, this is the same patch as in #295.

I just added these lines at the end of `memory-access.c:qmp_pmemaccess`:
~~~
   // wait for the socket to be available
    while (1) {
        if (access(path, F_OK) != -1)
            break;
        else
            // sleep 100 milliseconds
            usleep(100 * 1000);
    }

~~~
This way, this forces the socket to be created and binded when `qmp_pmemaccess` exits.
Otherwise there is a race condition where this thread has been created, but not scheduled, the command `pmemaccess` returns to the libvmi kvm driver, the driver tries to connect to the socket and it fails.

It's a little dirty hack which works for me now.
If you have better ideas, i'm listening.

Thanks !

### CODE REVIEW

**Comment by Wenzel** [comment]
Hi,
I updated the patch by remove the while loop and forcing the socket to be in a `listen` state before being passed to the thread, which will just `accept` new connections.
In this way we prevent a `connection refused` in the kvm driver because the thread was not scheduled.

The changes:
~~~C
struct pmemaccess_args {
    int socket_fd;
    struct sockaddr_un *address;
    socklen_t address_length;
    char *path;
    Error **errp;
};

static void *
memory_access_thread (void *p)
{
    int connection_fd;
    struct pmemaccess_args *pargs = (struct pmemaccess_args *)p;

    // accept incoming connections
    connection_fd = accept(pargs->socket_fd, (struct sockaddr *) pargs->address, &(pargs->address_length));
    connection_handler(connection_fd);

    close(pargs->socket_fd);
    unlink(pargs->path);
    free(pargs->path);
    free(pargs->address);
    free(pargs);
    return NULL;
}

void
qmp_pmemaccess (const char *path, Error **errp)
{
    pthread_t thread;
    sigset_t set, oldset;
    struct pmemaccess_args *pargs;

    // create the args struct
    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
    pargs->errp = errp;
    // create a copy of path that we can safely use
    pargs->path = malloc(strlen(path) + 1);
    memcpy(pargs->path, path, strlen(path) + 1);

    // create socket
    pargs->socket_fd = socket(PF_UNIX, SOCK_STREAM, 0);
    if (pargs->socket_fd < 0){
        error_setg(pargs->errp, "Qemu pmemaccess: socket failed");
        return NULL;
    }
    // unlink path if already exists
    unlink(path);
    // bind socket
    pargs->address = malloc(sizeof(struct sockaddr_un));
    pargs->address->sun_family = AF_UNIX;
    pargs->address_length = sizeof(pargs->address->sun_family) + sprintf(pargs->address->sun_path, "%s", (char *) pargs->path);
    if (bind(pargs->socket_fd, (struct sockaddr *) pargs->address, pargs->address_length) != 0){
        printf("could not bind\n");
        error_setg(pargs->errp, "Qemu pmemaccess: bind failed");
        return NULL;
    }

    // listen
    if (listen(pargs->socket_fd, 0) != 0){
        error_setg(pargs->errp, "Qemu pmemaccess: listen failed");
        return NULL;
    }

    // start the thread
    sigfillset(&set);
    pthread_sigmask(SIG_SETMASK, &set, &oldset);
    pthread_create(&thread, NULL, memory_access_thread, pargs);
    pthread_sigmask(SIG_SETMASK, &oldset, NULL);
}
~~~

**Comment by tklengyel** [comment]
This doesn't seem right to me: `memcpy(pargs->path, path, strlen(path) + 1);`

**Comment by carter-yagemann** [comment]
That line appears to assume that `path` is null terminated. Might want to error on the side of caution.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
++
++void
++qmp_pmemaccess (const char *path, Error **errp)
++{
++    pthread_t thread;
++    sigset_t set, oldset;
++    struct pmemaccess_args *pargs;
++
++    // create the args struct
++    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Check that pargs != NULL

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
++    pargs->errp = errp;
++    // create a copy of path that we can safely use
++    pargs->path = malloc(strlen(path) + 1);
++    memcpy(pargs->path, path, strlen(path) + 1);
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Don't copy past strlen(path) and manually null terminate pargs->path.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
++    // create socket
++    pargs->socket_fd = socket(PF_UNIX, SOCK_STREAM, 0);
++    if (pargs->socket_fd < 0){
++        error_setg(pargs->errp, "Qemu pmemaccess: socket failed");
++        return NULL;
++    }
++    // unlink path if already exists
++    unlink(path);
++    // bind socket
++    pargs->address = malloc(sizeof(struct sockaddr_un));
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Check that pargs->address is != NULL.

**Comment by Wenzel** [comment]
Changes highlight:
_Check that pargs != NULL_
~~~C
    // create the args struct
    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
    if (pargs == NULL){
        error_setg(errp, "Qemu pmemaccess: malloc failed");
        return NULL;
    }
~~~
_Don't copy past strlen(path) and manually null terminate pargs->path._
~~~C
    // create a copy of path that we can safely use
    size_t path_size = strlen(path);
    pargs->path = malloc(path_size + 1);
    memcpy(pargs->path, path, path_size);
    pargs->path[path_size] = '\0';
~~~

_Check that pargs->address is != NULL._
~~~C
    pargs->address = malloc(sizeof(struct sockaddr_un));
    if (pargs->address == NULL){
        error_setg(pargs->errp, "Qemu pmemaccess: malloc failed");
        return NULL;
    }
~~~

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
++    Error **errp;
++};
++
++#endif /* MEMORY_ACCESS_H */
+diff --git a/pixman b/pixman
+--- a/pixman
++++ b/pixman
+@@ -1 +1 @@
+-Subproject commit 87eea99e443b389c978cf37efc52788bf03a0ee0
++Subproject commit 87eea99e443b389c978cf37efc52788bf03a0ee0-dirty
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: This shouldn't be part of the patch. Results in:
```
File pixman is not a regular file -- refusing to patch
1 out of 1 hunk ignored -- saving rejects to file pixman.rej

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
++    if (bind(pargs->socket_fd, (struct sockaddr *) pargs->address, pargs->address_length) != 0){
++        printf("could not bind\n");
++        error_setg(pargs->errp, "Qemu pmemaccess: bind failed");
++        return NULL;
++    }
++
++    // listen
++    if (listen(pargs->socket_fd, 0) != 0){
++        error_setg(pargs->errp, "Qemu pmemaccess: listen failed");
++        return NULL;
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Returning NULL in a void function.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
++    if (pargs->address == NULL){
++        error_setg(pargs->errp, "Qemu pmemaccess: malloc failed");
++        return NULL;
++    }
++    pargs->address->sun_family = AF_UNIX;
++    pargs->address_length = sizeof(pargs->address->sun_family) + sprintf(pargs->address->sun_path, "%s", (char *) pargs->path);
++    if (bind(pargs->socket_fd, (struct sockaddr *) pargs->address, pargs->address_length) != 0){
++        printf("could not bind\n");
++        error_setg(pargs->errp, "Qemu pmemaccess: bind failed");
++        return NULL;
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Returning NULL in a void function.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
++        error_setg(pargs->errp, "Qemu pmemaccess: socket failed");
++        return NULL;
++    }
++    // unlink path if already exists
++    unlink(path);
++    // bind socket
++    pargs->address = malloc(sizeof(struct sockaddr_un));
++    if (pargs->address == NULL){
++        error_setg(pargs->errp, "Qemu pmemaccess: malloc failed");
++        return NULL;
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Returning NULL in a void function.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
++    size_t path_size = strlen(path);
++    pargs->path = malloc(path_size + 1);
++    memcpy(pargs->path, path, path_size);
++    pargs->path[path_size] = '\0';
++
++    // create socket
++    pargs->socket_fd = socket(PF_UNIX, SOCK_STREAM, 0);
++    if (pargs->socket_fd < 0){
++        error_setg(pargs->errp, "Qemu pmemaccess: socket failed");
++        return NULL;
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Returning NULL in a void function.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
++{
++    pthread_t thread;
++    sigset_t set, oldset;
++    struct pmemaccess_args *pargs;
++
++    // create the args struct
++    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
++    if (pargs == NULL){
++        error_setg(errp, "Qemu pmemaccess: malloc failed");
++        return NULL;
```
Comment (tklengyel) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Returning NULL in a void function.

**Comment by tklengyel** [comment]
This is OK to test.

**Comment by tklengyel** [comment]
I've installed QEMU 2.4.0.1 with this patch on the test system so we can actually try it live now.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by tklengyel** [comment]
If you squash the two commits into one, we can go ahead and merge this.
